### ISSUE TITLE
WINE configuration?

### ISSUE THREAD

**Issue by joepie91** [NixOS/nixpkgs#10165]
Not sure whether to file this on Nix or nixpkgs... Since _somebody_ around here is maintaining a WINE package, I figure I might as well ask it here :)

Anyhow, my usecase is wanting to build a Nix package for a Windows application, against a specific version and configuration of WINE, as well as a specific set of runtimes (think PlayOnLinux-style or winetricks, but preferably without either of those). How would one typically approach this with Nix?

I have no issues with reimplementing functionality for this if needed, I'm just not really sure what the packages would look like; the "bigger picture", so to say. My knowledge of Nix is still limited (especially where it concerns "plugin"-like things, like Windows runtimes under WINE), so I could use some direction here.

I should note that I don't necessarily need this to be merged into nixpkgs; hence my uncertainty where to file this issue.

**Comment by Ericson2314** [comment]
+1 To declaritively wrap a windows program with all the relevant wine settings would be amazing.

**Comment by cleverca22** [comment]
that does sound like it would be useful, but similar past issues&PR's have been shot down, #13625 #10006 #7597 #10981

**Comment by joepie91** [comment]
@cleverca22 To emphasize my previous statement:

> I should note that I don't necessarily need this to be merged into nixpkgs; hence my uncertainty where to file this issue.

It's not so much a proposal for modifying nixpkgs to include this functionality, it's more of a question on how to approach it from an express-writing point of view. I'll have a look at those PRs to understand the implementation, though - however, they don't seem to bring in any non-standard WINE versions or configurations from a quick glance.

**Comment by joelmo** [comment]
I would use buildEnv, you pass several different environments and makes symlinks to the passed derivations. But think about the registry, every wineprefix have it's own so one would probably need to merge these. You could create two different prefixs and install different apps to them to then see if any other files differs.

**Comment by Ericson2314** [comment]
I wasn't aware of that controversy, wow. Regardless, I see no issue with adding the infrastructure to declaratively create wine prefixes, as opposed to packaging of proprietary software with that infrastructure.

A good start would to make a branch with those rejected PRs, refactor out some common boilerplate, and get that merged but not the rejected packages.

**Comment by Profpatsch** [comment]
I think the question was answered, feel free to reopen otherwise.

**Comment by lopsided98** [comment]
I was looking in to implementing something like this, but I think it is infeasible in the general case. Most Linux software can separate its installation location from its configuration/data location, but this is not the case for most older Windows applications. They store configuration in their installation directory, and expect to be able to modify the registry at install and run time. Modern Windows software should store mutable data in AppData, but wine does not appear to provide a way to store this outside of the prefix, and you still have the problem of the registry.

The only way I could see of making declarative Wine prefixes work would be to use a filesystem overlay over the installation prefix in the Nix store to store mutable data. Windows actually does [something similar](https://blogs.msdn.microsoft.com/oldnewthing/20150902-00/?p=91681) to maintain compatibility with old software, but I think this is a lot of complexity for relatively little gain.

**Comment by deliciouslytyped** [comment]
Currently unusable WIP implementation experiment https://github.com/deliciouslytyped/nix-winnix

**Comment by fricklerhandwerk** [comment]
I also think offering infrastructure is a good thing. Not sure about complexity. Looks more like the approach is just opaque, and shedding some light onto it would suffice to get people going.

For a motivational example see [SoftEther](https://www.softether.org/) VPN, which is [available in nixpkgs](https://github.com/NixOS/nixpkgs/blob/f7574a5c8fefd86b50def1827eadb9b8cb266ffd/pkgs/servers/softether/4.29.nix), but only has a Windows GUI (which is refreshingly good, IMHO!) that you can run with `wine` to manage local or remote servers or clients.

For reference, in case somebody gets interested in this again:

Here is a mailing list conversation from 2017: [packaging windows applications through wine](https://releases.nixos.org/nix-dev/2017-April/023545.html) with [an example](https://github.com/openlab-aux/vuizvui/blob/97e440e996e9c418e46af3dbcdba58595b5c11ea/pkgs/aszlig/santander/default.nix#L63-L74) using `xdotool` to click installer buttons.

> Combination of bindfs+unionfs works well.
(bindfs for mangle permissions from 0444 to 0644, otherwise unionfs show
files from store as read-only. I picked unionfs as more bullet-proof,
don't know how kernel' overlayfs behave here)

@avnik do you have some sample code to show how exactly it works?

**Comment by lucasew** [comment]
I implemented a solution. The program often does not work from the readonly folder so my solution is basically a script that setups a wineprefix in a folder then install all the (wine)tricks on runtime in that prefix when first running.

I've tested with some games and packaged the best GUI compressor under the sun (7-Zip) in my personal overlay.

https://github.com/lucasew/nixcfg/blob/master/packages/wrapWine.nix

There are some examples in the folder wineApps
